File: Assets/Scripts/Core/Effects/ForceRevealAllEffect.cs
namespace Bunker.Core.Effects
{
    // Forces the target to reveal one random still-hidden trait to everyone.
    public class ForceRevealAllEffect : ISpecialCardEffect
    {
        public bool CanApply(GameSession session, PlayerData caster, PlayerData target)
        {
            if (target == null || target.IsEliminated) return false;
            if (target.PlayerId == caster.PlayerId) return false;
            return target.HasUnrevealedTraits();
        }

        public void Apply(GameSession session, PlayerData caster, PlayerData target)
        {
            // Picking which hidden trait gets forcibly revealed and firing the
            // public OnTraitRevealed event is handled by GameSession, since
            // it already owns that event and the random selection.
        }
    }
}

File: bunker-server/src/game/effects/ForceRevealAllEffect.ts
import { ISpecialCardEffect } from "./ISpecialCardEffect";
import { GameSession } from "../GameSession";
import { PlayerData } from "../PlayerData";

// Forces the target to reveal one random still-hidden trait to everyone.
export class ForceRevealAllEffect implements ISpecialCardEffect {
  canApply(_session: GameSession, caster: PlayerData, target: PlayerData | null): boolean {
    if (!target || target.isEliminated) return false;
    if (target.playerId === caster.playerId) return false;
    return target.hasUnrevealedTraits();
  }

  apply(_session: GameSession, _caster: PlayerData, _target: PlayerData | null): void {
    // No-op here — picking which hidden trait gets forcibly revealed and
    // firing the public traitRevealed event is handled by GameSession,
    // which already owns that event.
  }
}
